source/backend/dal/Repositories/ProjectRepository.cs

```diff
@@ -226,6 +231,51 @@ public void DeleteProjectDocument(long projectDocumentId)
             return;
         }
 
+        public PimsProject GetProjectAtTime(long projectId, DateTime time)
+        {
+            var projectHist = Context
+                .PimsProjectHists.AsNoTracking()
+                .Where(pacr => pacr.Id == projectId)
+                .Where(pacr => pacr.EffectiveDateHist <= time
+                    && (pacr.EndDateHist == null || pacr.EndDateHist > time))
+                .GroupBy(pacr => pacr.Id)
+                .Select(gpacr => gpacr.OrderByDescending(a => a.EffectiveDateHist).FirstOrDefault())
+                .FirstOrDefault();
+
+            var workActivityHist = Context
+                .PimsWorkActivityCodeHists.AsNoTracking()
+                .Where(pacr => pacr.Id == projectHist.WorkActivityCodeId)
+                .Where(pacr => pacr.EffectiveDateHist <= time
+                    && (pacr.EndDateHist == null || pacr.EndDateHist > time))
+                .GroupBy(pacr => pacr.Id)
+                .Select(gpacr => gpacr.OrderByDescending(a => a.EffectiveDateHist).FirstOrDefault())
+                .FirstOrDefault();
+
+            var costTypeHist = Context
+                .PimsCostTypeCodeHists.AsNoTracking()
+                .Where(pacr => pacr.Id == projectHist.CostTypeCodeId)
+                .Where(pacr => pacr.EffectiveDateHist <= time
+                    && (pacr.EndDateHist == null || pacr.EndDateHist > time))
+                .GroupBy(pacr => pacr.Id)
+                .Select(gpacr => gpacr.OrderByDescending(a => a.EffectiveDateHist).FirstOrDefault())
+                .FirstOrDefault();
+
+            var businessFunctionHist = Context
+                .PimsBusinessFunctionCodeHists.AsNoTracking()
+                .Where(pacr => pacr.Id == projectHist.BusinessFunctionCodeId)
+                .Where(pacr => pacr.EffectiveDateHist <= time
+                    && (pacr.EndDateHist == null || pacr.EndDateHist > time))
+                .GroupBy(pacr => pacr.Id)
+                .Select(gpacr => gpacr.OrderByDescending(a => a.EffectiveDateHist).FirstOrDefault())
+                .FirstOrDefault();
+
+            var project = _mapper.Map<PimsProject>(projectHist);
+            project.WorkActivityCode = _mapper.Map<PimsWorkActivityCode>(workActivityHist);
```

Comment:
Interesting, here you are doing this at the repo level, for compensation you were doing this at the service level. Why is that?

Reply:
Mostly to provide a similar api and return data from the non-historical version. On the comp-req I figured that the person and org historical models would be reused, so the corresponding repositories provide the methods.